colab-notebooks-examples/integrations/openai/intro.ipynb

```diff
@@ -0,0 +1,431 @@
+    "\n",
+    "We will first convert Pandas dataframe to helpers bulk api iteratable actions.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "efee9b97",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "\n",
+    "def dataframe_to_bulk_actions(df):\n",
+    "    for index, row in df.iterrows():\n",
+    "        yield {\n",
+    "            \"_index\": 'wikipedia_vector_index',\n",
+    "            \"_id\": row['id'],\n",
+    "            \"_source\": {\n",
+    "                'url' : row[\"url\"],\n",
+    "                'title' : row[\"title\"],\n",
+    "                'text' : row[\"text\"],\n",
+    "                'title_vector' : json.loads(row[\"title_vector\"]),\n",
+    "                'content_vector' : json.loads(row[\"content_vector\"]),\n",
+    "                'vector_id' : row[\"vector_id\"]\n",
+    "            }\n",
+    "        }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b8164b38",
+   "metadata": {},
+   "source": [
+    "As the dataframe is large, we will index data in batch of `100` and use python thread for faster indexing."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "aacb5e9c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from elasticsearch import helpers\n",
+    "\n",
+    "def batch_upload(start, end, batch_size):\n",
+    " for batch_start in range(start, end, batch_size):\n",
+    "    batch_end = min(batch_start + batch_size, len(wikipedia_dataframe))\n",
+    "    batch_dataframe = wikipedia_dataframe.iloc[batch_start:batch_end]\n",
+    "    actions = dataframe_to_bulk_actions(batch_dataframe)\n",
+    "    helpers.bulk(client, actions)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79083b01",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import threading\n",
+    "batch_size = 100\n",
+    "start = 0 \n",
+    "end = len(wikipedia_dataframe)\n",
+    "\n",
+    "thread1 = threading.Thread(target=batch_upload, args=(start, end//2, batch_size))\n",
+    "thread2 = threading.Thread(target=batch_upload, args=(end//2 , end, batch_size))\n",
+    "\n",
+    "thread1.start()\n",
+    "thread2.start()\n",
+    " \n",
+    "thread1.join()\n",
+    "thread2.join()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "091ffc51",
+   "metadata": {},
+   "source": [
+    "Let's test the index with simple match query."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2ccc8955",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(client.search(index=\"wikipedia_vector_index\", query={\n",
+    "    \"match\": {\n",
+    "      \"text\": {\n",
+    "        \"query\": \"Hummingbird\"\n",
+    "      }\n",
+    "    }\n",
+    "}))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "992b6804",
+   "metadata": {},
+   "source": [
+    "# Perform KNN search on dataset"
```

Comment:
Consider separating the question encoding and kNN search steps into distinct markdown headings :)

Reply:
Added in. a9a3d81a92ada6c8f53c09718937f0cdb5060fcc